lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1204,17 +1204,43 @@ Pattern *TypeChecker::coercePatternToType(ContextualPattern pattern,
   // TODO: permit implicit conversions?
   case PatternKind::Tuple: {
     TuplePattern *TP = cast<TuplePattern>(P);
-    bool hadError = type->hasError();
-    
+
+    // Allow optional promotion only for pattern matching, not pattern binding.
+    const bool allowOptionalPromotion = !pattern.getPatternBindingDecl();
+
+    // If optional promotion is allowed and the coercion type is optional,
+    // pretend we are in the context of an enum case payload.
+    if (allowOptionalPromotion && type->isOptional()) {
+      subOptions.setContext(TypeResolverContext::EnumPatternPayload);
+      subOptions |= TypeResolutionFlags::FromNonInferredPattern;
+    }
```

Comment:
Out of curiosity, why is this needed?

Reply:
I think I did this for correctness. These are the options used to coerce the subpattern in `case PatternKind::OptionalSome`.

Reply:
In that case, wouldn't they have been part of the `options`, not `subOptions`? Note though that `subOptions` has its context reset, so it wouldn't have been set to `TypeResolverContext::EnumPatternPayload` if the parent were an OptionalSomePattern, so I think we can just remove that (it only controls a ParenPattern -> TuplePattern transform when matching a single associated value enum with label, and that only matters at the top-level of an EnumElementPattern sub-pattern match, not within the elements of the tuple). Really that transform should be moved into the EnumElementPattern coercion logic anyways.

Regarding `TypeResolutionFlags::FromNonInferredPattern`, that's used to control the unintended inference diagnostic. Arguably because the OptionalSomePattern here is implicit, we shouldn't factor it into counting as an explicit pattern that changes how we infer types. So IMO we should just remove that too.

FWIW I'm hoping to eventually turn this into a visitor in CSApply so we don't have to pass around all these flags (https://github.com/apple/swift/commit/4bae220b849556fe144759c3dfe0f9c7e12334e4#diff-1ba7c144bdfaf2daf57703f890ef9575a5782850e77c573df4170a547d73ab99).

Reply:
> In that case, wouldn't they have been part of the `options`, not `subOptions`?

Ah, makes sense. And the rest as well, thanks for taking your time to explain these details.

> FWIW I'm hoping to eventually turn this into a visitor in CSApply so we don't have to pass around all these flags

Really looking forward to this! It‘d be great not to have to bother about type resolution here.